text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+### JavaScript Macro: dependencySatisfies
+
+Allows you to test if the given **allowed dependency** satisfies the given semver range. Always compiles to a boolean literal.
+
+```js
+import { dependencySatisfies, macroCondition } from "@ember/macros";
+if (macroCondition(dependencySatisfies("ember-data", "^3.0.0"))) {
+  // include code here for ember-data 3.0 compat
+}
+```
+
+The package version will be `semver.coerce`d first, such that non-standard versions like "3.9.0-beta.0" will appropriately satisfy constraints like "> 3.8".
+
+### Javascript Macro: failBuild
+
+Allow you to cause a build failure with a custom error message. If `failBuild` isn't eliminated by `macroCondition`'s branch elimination, the build will fail.
+
+```js
+import { dependencySatisfies, macroCondition, failBuild } from "@ember/macros";
+if (macroCondition(dependencySatisfies("ember-data", "^3.0.0"))) {
+  // include code here for ember-data 3.0 compat
+} else {
+  failBuild(`We don't support ember-data versions other than ^3.0.0`);
+}
+```
+
+### Handlebars Macro: macroGetOwnConfig
+
+`macroGetOwnConfig` is very similar to the `getOwnConfig` JS macro, but it works as a Handlebars helper. Given this `OwnConfig`:
+
+```json
+{
+  "items": [{ "score": 42 }]
+}
+```
+
+Then:
+
+```hbs
+<SomeComponent @score={{macroGetOwnConfig "items" 0 "score" }} />
+{{! ⬆️compiles to ⬇️ }}
+<SomeComponent @score={{42}} />
+```
+
+If you don't pass any arguments, you can get the whole thing (although this makes your template bigger, so use arguments when you can):
+
+```hbs
+<SomeComponent @config={{macroGetOwnConfig}} />
+{{! ⬆️compiles to ⬇️ }}
+<SomeComponent @config={{hash items=(array (hash score=42))}} />
+```
+
+### Handlebars Macro: macroCondition
+
+Used as a helper within a block `{{#if}}` or inline `{{if}}`. Just like the JS `macroCondition`, it ensures that branch elimination will happen.
+
+```hbs
+  {{#if (macroCondition (macroGetOwnConfig "shouldUseThing")) }}
+    <Thing />
+  {{else}}
+    <OtherThing />
+  {{/if}}
+
+  {{! ⬆️compiles to ⬇️ }}
+  <Thing />
+```
+
+### Handlebars Macro: macroDependencySatisfies
+
+Acts like a helper that returns a boolean. Like the `dependencySatisfies` JS macro, it can only resolve things that are **allowed dependencies**, so the same need for peer dependencies and/or **Optional Peer Dependencies** applies.
+
+```hbs
+<SomeComponent @canAnimate={{macroDependencySatisfies "liquid-fire" "*"}} />
+{{! ⬆️compiles to ⬇️ }}
+<SomeComponent @canAnimate={{true}} />
+```
+
+### Handlebars Macro: macroMaybeAttrs
+
+There is one place where `{{#if}}` doesn't work: within "element space". If you want to _sometimes_ set an attribute, but sometimes not, this doesn't work:
```

Comment:
Doesn't glimmer already treat `={{false}}` as "remove this attribute?

Reply:
What about modifiers?

Reply:
> Doesn't glimmer already treat ={{false}} as "remove this attribute?

Does it do that as a compile-time optimization? If so, then that's great and we can rely on that.

> What about modifiers?

That is a good question, perhaps we still have a small hole here. I came up with `maybeMacroAttrs` because it was a clear case that popular addons are doing today. I'm not aware of any that conditionally remove a modifier invocation, but that seems plausible.

Reply:
> Does it do that as a compile-time optimization? If so, then that's great and we can rely on that.

I don't think it does, but assuming I was correct, and that this behavior is universal (i.e. not conditional on "oh let's see if we are doing an attribute vs property, etc" then I don't see why not). But even if it's not optimized yet, if that behaves correctly (i.e. not adding the attribute to the DOM), then I don't think `maybeMacroAttrs` is necessary.